server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -170,6 +171,19 @@ public CoordinationDiagnosticsService(
         this.nodeHasMasterLookupTimeframe = NODE_HAS_MASTER_LOOKUP_TIMEFRAME_SETTING.get(clusterService.getSettings());
         this.unacceptableNullTransitions = NO_MASTER_TRANSITIONS_THRESHOLD_SETTING.get(clusterService.getSettings());
         this.unacceptableIdentityChanges = IDENTITY_CHANGES_THRESHOLD_SETTING.get(clusterService.getSettings());
+        /*
+         * This is called here to cover an edge case -- when there are master-eligible nodes in the cluster but none of them has been
+         * elected master. In the most common case this node will receive a ClusterChangedEvent that results in this polling being
+         * cancelled almost immediately. If that does not happen, then we do in fact need to be polling. Unfortunately there is no way to
+         * tell at this point whether this node is master-eligible or not, so we kick this off regardless. On master-eligible nodes the
+         * results will always be harmlessly ignored. Note that beginPollingRemoteMasterStabilityDiagnostic results in several internal
+         * transport actions being called, so it must run in the system context.
+         */
+        final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
+        try (ThreadContext.StoredContext ignored = threadContext.stashContext()) {
+            threadContext.markAsSystemContext();
+            beginPollingRemoteMasterStabilityDiagnostic();
+        }
```

Comment:
Doing business logic in the constructor is an anti-pattern as you're racing against the initializer thread (e.g. an uncomplete`this` could escape in the async code)

I think  the `clusterService.addListener(this);` calls should be done outside the constructor too (as it publishes `this`)

Should we have an `init` or `start` method where we do these things and call this new method from the outside after we called the constructor?

Reply:
Yeah I can do that.

Reply:
Actually by doing that I can call init() much later in Node's startup, and avoid several of the other problems in this PR.